README.md
# mseviz
- Version: 0.2.5
- Date: 28-10-2021
- Author: Iago Mosqueira, EC JRC.
- Maintainer: Iago Mosqueira, EC JRC.
- Repository: <https://github.com/iagomosqueira/mseviz/>
- Bug reports: <https://github.com/iagomosqueira/mseviz/issues>

## Overview
Plots and Visualization Tools for Management Strategy Evaluation Results

To install this package, start R and enter:

  library(devtools)
  install_github("iagomosqueira/mseviz")

## Documentation
- Help pages
- Vignette

## Build Status
[![Travis Build Status](https://travis-ci.org/iagomosqueira/mseviz.svg?branch=master)](https://travis-ci.org/iagomosqueira/mseviz)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/iagomosqueira/mseviz?branch=master&svg=true)](https://ci.appveyor.com/project/iagomosqueira/mseviz)

## Releases
- [All releases](https://github.com/iagomosqueira/mseviz/releases/)

## License
Copyright (c) 2017-18 European Union. European Commission Joint Research Centre D.02. Released under the [EUPL 1.1](https://joinup.ec.europa.eu/community/eupl/og_page/eupl).

## Contact
You are welcome to:

- Submit suggestions and bug-reports at: <https://github.com/iagomosqueira/mseviz/issues>
- Send a pull request on: <https://github.com/iagomosqueira/mseviz/>
- Compose a friendly e-mail to: <iago.mosqueira@ec.europa.eu>


NEWS
 mseviz 0.2.5

 BUG FIXES

- vignette uses correct names for statistics

 USER-VISIBLE CHANGES

- Moved from indicators to statistics, following package:mse name change.


tests/plots.R
# plots.R - DESC
# /plots.R

# Copyright European Union, 2017
# Author: Iago Mosqueira (EC JRC) <iago.mosqueira@ec.europa.eu>
#
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

library(mseviz)

# plotTOs

data(perf)

# 1. Boxplots

plotBPs(perf)
plotBPs(perf, target=c(S3=1))

# 2. Trade-offs

plotTOs(perf, "Y1", c("S3", "S5", "T1", "S6"))




R/tables.R
# tables.R - DESC
# /tables.R

# Copyright European Union, 2018
# Author: Iago Mosqueira (EC JRC) <iago.mosqueira@ec.europa.eu>
#
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

# TODO final table
# mean(performance) for 1, 5, 10, 20 years

# TODO initial table 1
# mean(perf) by indicator
# - CONVERT 0 to < 0.01
# - CONVERT 1 to > 0.99

foo <- function(expr) {

 # CHANGE [] to  _{}
 expr <- gsub("[", "_{", expr, fixed=TRUE)
 expr <- gsub("]", "}", expr, fixed=TRUE)
 expr <- gsub("hat\\((.*)\\)", "\\\\hat{\\1}", expr)
 # expr <- gsub("<=", "\\\\leq", expr, fixed=TRUE)
 # expr <- gsub(">=", "\\\\geq", expr, fixed=TRUE)
 # expr <- gsub("%*%", "\\\\dot", expr, fixed=TRUE)
 return(paste0("$", expr, "$"))
}

# summTable {{{

summTable <- function(data,  indicators=unique(data[['indicator']]),
  probs=c(0.10, 0.50, 0.90), ...) {

  # SUBSET indicators
  data <- data[indicator %in% indicators,]

  # CONVERT name to LaTeX
  data[, name:=foo(name)]

  # CALCULATE quantiles
  qdata <- data[,as.list(quantile(data, probs=probs, na.rm=TRUE)),
    keyby=list(indicator, name, mp)]
  
  qdata[, fig:=paste0(
    format(round(`50%`, 2), digits=1, scientific=FALSE, trim=TRUE), " (",
    format(round(`10%`, 2), digits=1, trim=TRUE, scientific=FALSE), "-",
    format(round(`90%`, 2), digits=1, trim=TRUE, scientific=FALSE), ")")]
 
  qtab <- dcast(qdata, mp ~ name, value.var = "fig")
  
  # CALCULATE means
  mdata <- data[, .(fig=format(mean(data), digits=2, scientific=FALSE, trim=TRUE)),
    keyby=list(indicator, name, mp)]
  
  mtab <- dcast(mdata, mp ~ name, value.var = "fig")
  
  # ASSEMBLE table
  tab <- cbind(qtab[,1], qtab[,4], mtab[,2], mtab[,3], qtab[,6], mtab[,5])
  
  # RANK by row
  qrank <- dcast(qdata, mp ~ name, value.var = "50%")
  cols <- colnames(qrank)[-1]
  qrank[, (cols) := lapply(.SD, frank, ties.method="first"), .SDcols = cols]
  qrank <- qrank[, colnames(tab), with=FALSE]
  
  # COLOUR by rank
  colours <- c(rep("000000", 2), rep("343434", 4), rep("665544", 2))
  
  # OUTPUT table
  return(xtable(tab, ...))

} # }}}

# resTable {{{
# data <- perft[year == 2023]
# desc <- Reduce(rbind,lapply(indicators, function(x) as.data.frame(x[2:3])))

resTable <- function(data, indicators=unique(data[['indicator']]), ...) {

  desc <- Reduce(rbind, lapply(indicators, function(x) as.data.frame(x[2:3])))

  # COMPUTE mean by indicator & mp
  mdat <- data[, .(data=mean(data)), by=.(indicator, mp, name)]

  # MERGE indicators description
  tdat <- merge(mdat, desc, by.x='name', by.y='name')
  
  tab <- dcast(tdat, desc + name ~ mp, value.var = "data")

  # CONVERT name to LaTeX
  tab[, name:=foo(name)]
  
  # FORMAT zeroes
  # tab[tab == 0] <- NA
  
  xtable(tab, ...)
} # }}}


data-raw/runsom.R
# runsom.R - DESC
# /runsom.R

# Copyright European Union, 2018
# Author: Iago Mosqueira (EC JRC) <iago.mosqueira@ec.europa.eu>
#
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

library(FLasher)
library(FLBRP)

# DATA
data(ple4)
data(ple4.indices)

# SR and BRP
p4sr <- fmle(as.FLSR(ple4, model='ricker'))
p4rp <- brp(FLBRP(ple4, sr=p4sr))

# OMP
omp <- propagate(fwdWindow(ple4, p4rp, end=2038)[, ac(2008:2038)], 250)
residuals <- rlnorm(250, residuals(p4sr), 0.5)[,sample(1:51, 30)]
dimnames(residuals) <- list(year=2009:2038)

# fwd(omp)
omps <- FLStocks(parallel::mclapply(seq(0.1, 0.4, length=6), function(x) {
  fwd(omp, sr=p4sr, residuals=residuals,
    control=fwdControl(year=2009:2038, quant='f', value=runif(250*30, x, x*1.5)))
  }, mc.cores = 2))

# RUNS
names(omps) <- paste0("MP", 1:6)

library(data.table)

# METRICS
runs <- lapply(omps, metrics, metrics=list(F=fbar, SB=ssb, C=catch))
# runs <- rbindlist(lapply(runs, as.data.frame, drop=TRUE), idcol="mp")

# OM
om <- metrics(ple4, metrics=list(F=fbar, SB=ssb, C=catch))
#om <- data.table(as.data.frame(metrics(ple4,
#  metrics=list(F=fbar, SB=ssb, C=catch)), drop=TRUE))
# om[, iter:=1]
#om <- om[,c(1,4,2,3)]

save(om, runs, file="../data/omruns.RData")


# PERF

library(mse)

data(statistics)

perf <- performance(omps, metrics=list(F=fbar, SB=ssb, C=catch),
  refpts=FLPar(SB0=1500000, SBMSY=400000, Ftarget=0.21, FMSY=0.21, SBlim=200000),
  statistics=statistics[c(1,2,4,5,6,9,10,11,12)], years=list(2020:2038))

save(perf, file="../data/perf.RData")



R/plots.R
# plots.R - DESC
# ioalbmse/R/plots.R

# Copyright European Union, 2015-2016
# Author: Iago Mosqueira (EC JRC) <iago.mosqueira@ec.europa.eu>
#
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

# plotBPs {{{

#' Boxplot by MP for a range of statistics
#' Figure 3
#' @examples
#' data(perf)
#' # A data.table of performance statistics per run,
#' head(perf)
#' # plot selected statistics
#' plotBPs(perf, statistics=c("SB0", "FMSY", "green"))
#' # Add targets and limits by statistics, as named vectors
#' plotBPs(perf, statistics=c("SB0", "FMSY", "green"),
#'   target=c(SB0=0.40, FMSY=1, green=0.5), limit=c(SB0=0.10))
#' # size control the diameter of the point behind thin boxplots
#' plotBPs(perf, statistics=c("SB0", "FMSY", "green"), size=3)

plotBPs <- function(data, statistics=unique(data$statistic), size=3,
  target=missing, limit=missing, yminmax=c(0.10, 0.90), lowupp=c(0.25, 0.75)) {

  # CHECK quantiles
  if(any(c(length(yminmax), length(lowupp)) != 2))
    stop("'yminmax' and 'lowupp' must be both of length 2")

  # SUBSET statistics
  data <- data[statistic %in% statistics,]
  
  # ORDER name as of statistics
  cols <- unique(data[,c('statistic','name')])
  data[, name:=factor(name, levels=cols$name[match(cols$statistic, statistics)],
    ordered=TRUE)]
  
  # ORDER mp as in input
  data[, mp:=factor(mp, levels=unique(mp))]

  dat <- data[, .(ymin=quantile(data, yminmax[1], na.rm=TRUE),
    lower=quantile(data, lowupp[1], na.rm=TRUE),
    middle=median(data, na.rm=TRUE), upper=quantile(data, lowupp[2], na.rm=TRUE),
    ymax=quantile(data, yminmax[2], na.rm=TRUE)), by=.(mp, statistic, name)]

  # PLOT
  p <- ggplot(dat,
    aes(x=mp, ymin=ymin, lower=lower, middle=middle,
      upper=upper, ymax=ymax, fill=mp)) +
    # data ~ mp, colour by mp
    # PLOT point by mp, useful if boxplot is very thin
    geom_point(data=dat[, .(middle=mean(middle)), by=.(mp, name)],
      aes(x=mp, y=middle), colour="black", size=size + size*0.20, inherit.aes=FALSE) +
    geom_point(data=dat[, .(middle=mean(middle)), by=.(mp, name)],
      aes(x=mp, y=middle, colour=mp), size=size, inherit.aes=FALSE) +
    # PLOT boxplot by mp
    geom_boxplot(stat="identity") +
    # PANELS per statistics
    facet_wrap(~name, scales='free_y') +
    xlab("") + ylab("") +
    # DELETE x-axis labels, LEGEND in 6th panel
    # TODO legend pos by no. of panels
    theme(axis.text.x=element_blank(), legend.position=c("right"),
    # DELETE legend title
    legend.title=element_blank())

  # TARGET
  if(!missing(target)) {
    dat <- data[statistic %in% names(target),]
    dat[, target:=unlist(target)[match(statistic, names(target))]]
    p <- p + geom_hline(data=dat, aes(yintercept=target), colour="green",
      linetype="longdash", size=1)
  }
  
  # LIMIT
  if(!missing(limit)) {
    dat <- data[statistic %in% names(limit),]
    dat[, limit:=unlist(limit)[match(statistic, names(limit))]]
    p <- p + geom_hline(data=dat, aes(yintercept=limit), colour="red",
      linetype="longdash", size=1)
  }

  return(p)
} # }}}

# plotTOs {{{

#' Trade-offs plot by MP for a range of statistics
#' Figure 4
#' @examples
#' data(perf)
#' plotTOs(perf, x="C", y=c("SBMSY", "FMSY", "green", "SB0"))

plotTOs <- function(data, x=unique(data$statistic)[1],
  y=setdiff(unique(data$statistic), x), probs=c(0.10, 0.50, 0.90),
   size=0.50, alpha=0.75) {
  
  # ORDER mp as in input
  data[, mp:=factor(mp, levels=unique(mp))]

  # CALCULATE quantiles
  data <- data[, as.list(quantile(data, probs=probs, na.rm=TRUE)),
    keyby=list(statistic, name, year, mp)]

  # LABELS probs
  xsyms <- syms(paste0("x", probs))
  ysyms <- syms(paste0("y", probs))

  # SUBSET statistics
  daty <- data[statistic %in% y,]
  setnames(daty, seq(5, 4 + length(probs)), paste0("y", probs))
  datx <- data[statistic %in% x,]
  setnames(datx, seq(5, 4 + length(probs)), paste0("x", probs))
  
  # MERGE x into y
  dat <- cbind(daty, datx[,-(1:4)])
 
  p <- ggplot(dat, aes(x=!!xsyms[[2]], y=!!ysyms[[2]])) +
    xlab(unique(datx$name)) + ylab("") +
  # PLOT lines
  geom_linerange(aes(ymin=!!ysyms[[1]], ymax=!!ysyms[[3]]), size=size, alpha=alpha) +
  geom_linerange(aes(xmin=!!xsyms[[1]], xmax=!!xsyms[[3]]), size=size, alpha=alpha) +
  # PLOT median dots
  geom_point(aes(fill=mp), shape=21, size=4) +
  facet_wrap(~name, scales="free_y") +
  scale_shape(solid=FALSE) + theme(legend.title=element_blank())

  return(p)
}
# }}}

# kobeMPs {{{

#' @examples
#' kobeMPs(perf)

kobeMPs <- function(data, x="SBMSY", y="FMSY", xlim=0.40, ylim=1.4,
  probs=c(0.10, 0.50, 0.90), size=0.75, alpha=1) {
  
  # CALCULATE quantiles
  data <- data[, as.list(quantile(data, probs=probs, na.rm=TRUE)),
    keyby=list(statistic, name, year, mp)]

  # SUBSET statistics
  daty <- data[statistic %in% y,]
  setnames(daty, seq(5, 4 + length(probs)), paste0("y", paste0(probs*100, "%")))
  datx <- data[statistic %in% x,]
  setnames(datx, seq(5, 4 + length(probs)), paste0("x", paste0(probs*100, "%")))
  
  # MERGE x into y
  data <- cbind(daty, datx[,-(1:4)])
  lim <- pmax(round(apply(data[,c('y90%','x90%')] / 0.5, 2, max)) * 0.5, 1.5)

  # PLOT
  p <- ggplot(data, aes(x=`x50%`, y=`y50%`)) +
    # Kobe background
    geom_rect(aes(xmin=1, xmax=Inf, ymin=0, ymax=1), colour='green',
      fill='green') +
    geom_rect(aes(xmin=0, xmax=1, ymin=0, ymax=1), colour='yellow',
      fill='yellow') +
    geom_rect(aes(xmin=1, xmax=Inf, ymin=1, ymax=Inf), colour='orange', 
      fill='orange') +
    geom_rect(aes(xmin=0, xmax=1, ymin=1, ymax=Inf), colour='red',
      fill='red') +
    # Central GRID
    geom_hline(aes(yintercept=1)) + geom_vline(aes(xintercept=1)) +
    # lims
    scale_x_continuous(expand = c(0, 0), limits = c(0, lim['x90%'])) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, lim['y90%'])) +
    # DROP background grid
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
      panel.background = element_blank(), axis.line = element_blank()) +
    # PLOT LRPs
    geom_segment(aes(x=xlim, xend=Inf, y=ylim, yend=ylim), colour='gray') +
    geom_segment(aes(x=xlim, xend=xlim, y=0, yend=ylim), colour='gray') +
    # PLOT lines
    geom_linerange(aes(ymin=`y10%`, ymax=`y90%`), size=size, alpha=alpha) +
    geom_linerange(aes(xmin=`x10%`, xmax=`x90%`), size=size, alpha=alpha) +
    # PLOT median dots
    geom_point(aes(fill=mp), shape=21, size=4) +
    scale_shape(solid=FALSE) + theme(legend.title=element_blank()) +
    # LABELS
    labs(x=expression(SB/SB[MSY]), y=expression(F/F[MSY])) +
    annotate("text", x = xlim - xlim * 0.35, y = 0.1,
      label = "SB[lim]", parse=TRUE) +
    annotate("text", x = 0.90, y = 0.1, label = "SB[targ]", parse=TRUE) +
    annotate("text", x = lim["x90%"] - lim["x90%"] * 0.10,
      y = ylim + ylim * 0.10,
      label = "F[lim]", parse=TRUE) +
    annotate("text", x =  lim["x90%"] - lim["x90%"] * 0.10, y = 1.10,
      label = "F[targ]", parse=TRUE)

  return(p)
} # }}}

# kobeTS {{{

kobeTS <- function(perfts) {

  ggplot(perfts, aes(x=ISOdate(year, 1, 1), y=data, fill=statistic)) + 
    geom_col(colour="black", size=0.5) +
    scale_discrete_manual(name = "Kobe Quadrant", aesthetics=c("fill"),
      values=c(green="darkgreen", red="red", yellow="yellow2", orange="orange")) +
    facet_wrap(~mp) +
    xlab("") + ylab("") +
    theme(legend.position=c(.85,.15))

} # }}}

# plotOMruns {{{

plotOMruns <- function(om, runs, limit=missing, target=missing, iter=NULL,
  probs=c(0.10, 0.25, 0.50, 0.75, 0.90), iyear=dims(om)$maxyear, ylab="", ylim=missing) {

  # CHECK classses
  if(!is(om, "FLQuant") | !is(runs, "FLQuants"))
    stop("om and runs must be of class FLQuant and FLQuants respectively.")
 
  # PLOT om
  p1 <- ggplotFL::plot(om, probs=probs) + xlim(NA, iyear + 1) +
    geom_vline(xintercept=iyear)

  # RPs
  if(!missing(limit))
    p1 <- p1 + geom_hline(aes(yintercept=limit), colour="red", linetype=2)
  if(!missing(target))
    p1 <- p1 + geom_hline(aes(yintercept=target), colour="green", linetype=2)

  p2 <- ggplotFL::plot(runs, probs=probs, iter=iter) + facet_wrap(~qname,
    ncol=2) + ylab(ylab) + geom_vline(xintercept=iyear)

  # RPs
  if(!missing(limit))
    p2 <- p2 + geom_hline(aes(yintercept=limit), colour="red", linetype=2)
  if(!missing(target))
    p2 <- p2 + geom_hline(aes(yintercept=target), colour="green", linetype=2)

  if(!missing(ylim))
    p2 <- p2 + coord_cartesian(ylim=ylim)
  
  p <- p1 + p2 + plot_layout(ncol=1, heights=c(1, length(runs) / 2))

  return(p)

} # }}}
